File: collectors/indeed/collect.py
#!/usr/bin/env python3
"""
Collects Indeed postings via `python-jobspy` and pushes them to
argos-career's authenticated ingest endpoint.

ADR-027 (why this runs outside argos-career's own container — no Docker
socket mounted there, this is a fully separate process) and ADR-028 (why
this is allowed to break CLAUDE.md's polite-collector rule for this one
library and this one host — jobspy's Indeed scraper cannot be configured to
use an honest User-Agent, and apis.indeed.com's robots.txt disallows
everything regardless).

One run, one exit — this is the "ephemeral container" `docker run --rm`
runs on a schedule (see argos-indeed-collect.timer), not a long-lived
process. Never on the critical path: a failure here means Indeed's
postings are stale until the next scheduled run, nothing else in the
pipeline depends on it succeeding (principle 1, extended to a source that
lives outside the app entirely).

Required environment:
  ARGOS_API_URL   e.g. http://100.x.x.x:3000 (Atlas's Tailscale address)
  ARGOS_API_KEY   the same Bearer key every other API caller uses

Optional environment (defaults below):
  SEARCH_TERM, LOCATION, COUNTRY_INDEED, RESULTS_WANTED
"""

import json
import os
import sys

import requests
from jobspy import scrape_jobs

DEFAULT_SEARCH_TERM = "estagio"
DEFAULT_LOCATION = "Rio de Janeiro, Brazil"
DEFAULT_COUNTRY_INDEED = "Brazil"
DEFAULT_RESULTS_WANTED = "50"


def env(name: str, default: str | None = None, required: bool = False) -> str:
    value = os.environ.get(name, default)
    if required and not value:
        print(f"ERROR: {name} is required", file=sys.stderr)
        sys.exit(1)
    return value or ""


def main() -> None:
    search_term = env("SEARCH_TERM", DEFAULT_SEARCH_TERM)
    location = env("LOCATION", DEFAULT_LOCATION)
    country_indeed = env("COUNTRY_INDEED", DEFAULT_COUNTRY_INDEED)
    results_wanted = int(env("RESULTS_WANTED", DEFAULT_RESULTS_WANTED))
    api_url = env("ARGOS_API_URL", required=True).rstrip("/")
    api_key = env("ARGOS_API_KEY", required=True)

    print(
        f"jobspy: searching Indeed for '{search_term}' in '{location}' "
        f"(up to {results_wanted})"
    )
    jobs = scrape_jobs(
        site_name=["indeed"],
        search_term=search_term,
        location=location,
        country_indeed=country_indeed,
        results_wanted=results_wanted,
    )
    print(f"jobspy: {len(jobs)} rows returned")

    if len(jobs) == 0:
        print("nothing to ingest, exiting")
        return

    # jobspy returns a pandas DataFrame; round-tripping through its own
    # to_json is what turns NaN into proper JSON null (a bare json.dumps on
    # the DataFrame's dict form does not) — the same conversion used to
    # capture the real fixture this source's schema/normalizer were fitted
    # against (ADR-027).
    rows = json.loads(jobs.to_json(orient="records", date_format="iso"))

    # jobspy's own row id is the natural sourceId — stable per posting, the
    # same field the normalizer's schema requires.
    postings = [
        {"sourceId": row["id"], "payload": row} for row in rows if row.get("id")
    ]
    skipped = len(rows) - len(postings)
    if skipped:
        print(f"WARNING: {skipped} row(s) had no id, skipped")

    # jobspy has no "there were more, we stopped" signal of its own -- a
    # result count that reached the requested budget is the same heuristic
    # this project's other paginated collectors use (a full final page plus
    # a cap being what stopped it, not the source running dry), applied
    # here since this process never sees Indeed's raw response either
    # (docs/audit PR-015).
    truncated = len(jobs) >= results_wanted

    body = {"source": "indeed", "postings": postings, "truncated": truncated}
    response = requests.post(
        f"{api_url}/runs/collect/external",
        json=body,
        headers={"Authorization": f"Bearer {api_key}"},
        timeout=120,
    )
    print(f"ingest: HTTP {response.status_code}")
    print(response.text[:2000])
    response.raise_for_status()


if __name__ == "__main__":
    main()

File: collectors/catho/collect.ts
/**
 * Collects Catho postings via a real headless browser and pushes them to
 * argos-career's authenticated ingest endpoint.
 *
 * See docs/adr/032-catho-collector-headless-browser.md for the full
 * discovery record. Unlike Indeed (ADR-028), this collector does NOT break
 * CLAUDE.md §6's polite-collector rules: `robots.txt` only disallows
 * Catho's *search* path (`/buscar/vagas/`), which this script never
 * touches, and Playwright's real Chromium User-Agent is an honest
 * statement of what it is, not a forged one — individual posting pages
 * (`/vagas/<slug>/<id>/`) return a plain 403 to a non-browser UA but are
 * not disallowed by robots.txt at all.
 *
 * The one real cost this shape has that Gupy/Sólides don't: there is no
 * server-side search, so there is no way to ask Catho for only
 * Rio-de-Janeiro-metro internships. Every candidate posting (title-filtered
 * from the sitemap) has to be opened once to learn its real city. Bounded
 * per run (MAX_PAGES_PER_RUN) rather than run to completion in one process —
 * a state file tracks which posting IDs have already been visited
 * (ingested or found expired) so repeated runs make incremental progress
 * through the backlog instead of re-walking it, and so a normal run, once
 * the backlog is drained, only touches what's actually new.
 *
 * One run, one exit — matches the "ephemeral container" shape
 * `collectors/indeed` already established. Never on the critical path: a
 * failure here means Catho's postings are stale until the next scheduled
 * run (principle 1, extended to a source that lives outside the app
 * entirely).
 *
 * Required environment:
 *   ARGOS_API_URL   e.g. http://100.x.x.x:3000 (Atlas's Tailscale address)
 *   ARGOS_API_KEY   the same Bearer key every other API caller uses
 *
 * Optional environment (defaults below):
 *   MAX_PAGES_PER_RUN, REQUEST_INTERVAL_MS, SEEN_IDS_PATH, TITLE_PATTERN
 */
import { chromium } from "playwright";
import { readFileSync, writeFileSync, existsSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";

const SITEMAP_INDEX = "https://www.catho.com.br/sitemap-index.xml";
// Only the fresh, numbered "sitemap_vagas_N.xml" set (regenerated daily,
// confirmed via discovery) — not the legacy `sitemap_vagas_emprego.xml`
// (dated 2011, company-profile URLs, not postings) and not the gzipped
// `busca-vagas/sitemapN.xml.gz` set (unconfirmed content, not fetched
// during discovery — revisit if the numbered set ever stops being enough).
const SITEMAP_ENTRY_PATTERN = /\/sitemap2\/sitemap_vagas_\d+\.xml$/;

const DEFAULT_TITLE_PATTERN =
  "estagio|estagiario|estagiaria|estágio|estagiário|estagiária|trainee";
const DEFAULT_MAX_PAGES_PER_RUN = 300;
const DEFAULT_REQUEST_INTERVAL_MS = 1_500;
const DEFAULT_SEEN_IDS_PATH = "/data/catho-seen-ids.json";
const FETCH_UA =
  "ArgosCareer/0.1.0 (+https://github.com/gustavopinto244/ArgosCareer; personal internship search bot)";

function env(name: string, fallback: string): string {
  return process.env[name] ?? fallback;
}

function requiredEnv(name: string): string {
  const value = process.env[name];
  if (!value) {
    console.error(`ERROR: ${name} is required`);
    process.exit(1);
  }
  return value;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

interface SitemapCandidate {
  id: string;
  url: string;
}

async function fetchText(url: string): Promise<string> {
  const response = await fetch(url, { headers: { "User-Agent": FETCH_UA } });
  if (!response.ok) {
    throw new Error(`${url} responded ${response.status}`);
  }
  return response.text();
}

function extractLocs(xml: string): string[] {
  return [...xml.matchAll(/<loc>([^<]+)<\/loc>/g)].map((m) => m[1]!.trim());
}

/**
 * A candidate's stable ID is the numeric segment at the end of its sitemap
 * URL (`/vagas/<slug>/<id>/`) — independent of Catho's own internal "offer"
 * ID scheme, which the page never exposes to a plain sitemap read.
 */
function toCandidate(url: string): SitemapCandidate | null {
  const match = /\/vagas\/[^/]+\/(\d+)\/?$/.exec(url);
  return match ? { id: match[1]!, url } : null;
}

async function discoverCandidates(
  titlePattern: RegExp,
): Promise<SitemapCandidate[]> {
  const indexXml = await fetchText(SITEMAP_INDEX);
  const sitemapUrls = extractLocs(indexXml).filter((url) =>
    SITEMAP_ENTRY_PATTERN.test(url),
  );
  console.log(`sitemap index: ${sitemapUrls.length} vaga sitemap(s) found`);

  const candidates: SitemapCandidate[] = [];
  for (const sitemapUrl of sitemapUrls) {
    const xml = await fetchText(sitemapUrl);
    const urls = extractLocs(xml);
    for (const url of urls) {
      if (!titlePattern.test(url)) continue;
      const candidate = toCandidate(url);
      if (candidate) candidates.push(candidate);
    }
    console.log(`${sitemapUrl}: ${urls.length} urls scanned`);
  }
  return candidates;
}

function loadSeenIds(path: string): Set<string> {
  if (!existsSync(path)) return new Set();
  try {
    const raw: unknown = JSON.parse(readFileSync(path, "utf8"));
    return new Set(Array.isArray(raw) ? raw.map(String) : []);
  } catch {
    console.warn(
      `WARNING: could not parse ${path}, starting with an empty seen-set`,
    );
    return new Set();
  }
}

function saveSeenIds(path: string, ids: Set<string>): void {
  mkdirSync(dirname(path), { recursive: true });
  writeFileSync(path, JSON.stringify([...ids].sort()), "utf8");
}

interface CollectedPosting {
  sourceId: string;
  payload: {
    id: string;
    url: string;
    pageTitle: string;
    jobPosting: unknown;
  };
}

/**
 * Opens one posting page with a real Chromium browser and extracts its
 * `application/ld+json` `JobPosting` markup and page `<title>`.
 *
 * Returns `null` — not a throw — when the page redirects away from its own
 * URL (an expired posting falling back to the generic listing, confirmed
 * during discovery) or carries no JSON-LD at all. The caller still marks
 * the ID as seen either way: revisiting a confirmed-expired posting on the
 * next run wastes a page load for no gain.
 */
async function collectOne(
  page: import("playwright").Page,
  candidate: SitemapCandidate,
): Promise<CollectedPosting | null> {
  const response = await page.goto(candidate.url, {
    waitUntil: "domcontentloaded",
    timeout: 20_000,
  });
  if (!response || !response.ok()) return null;

  const finalUrl = page.url();
  if (!finalUrl.includes(`/${candidate.id}`)) {
    // Redirected away from its own posting path — expired, per discovery.
    return null;
  }

  const jobPosting = await page.evaluate(() => {
    const script = document.querySelector('script[type="application/ld+json"]');
    if (!script?.textContent) return null;
    try {
      return JSON.parse(script.textContent) as unknown;
    } catch {
      return null;
    }
  });
  if (!jobPosting) return null;

  const pageTitle = await page.title();

  return {
    sourceId: candidate.id,
    payload: { id: candidate.id, url: candidate.url, pageTitle, jobPosting },
  };
}

async function main(): Promise<void> {
  const titlePattern = new RegExp(
    env("TITLE_PATTERN", DEFAULT_TITLE_PATTERN),
    "i",
  );
  const maxPagesPerRun = Number(
    env("MAX_PAGES_PER_RUN", String(DEFAULT_MAX_PAGES_PER_RUN)),
  );
  const requestIntervalMs = Number(
    env("REQUEST_INTERVAL_MS", String(DEFAULT_REQUEST_INTERVAL_MS)),
  );
  const seenIdsPath = env("SEEN_IDS_PATH", DEFAULT_SEEN_IDS_PATH);
  const apiUrl = requiredEnv("ARGOS_API_URL").replace(/\/$/, "");
  const apiKey = requiredEnv("ARGOS_API_KEY");

  const seenIds = loadSeenIds(seenIdsPath);
  console.log(`seen-ids: ${seenIds.size} known from previous runs`);

  const candidates = await discoverCandidates(titlePattern);
  const unseen = candidates.filter((c) => !seenIds.has(c.id));
  console.log(
    `candidates: ${candidates.length} title-matched, ${unseen.length} not yet seen`,
  );

  const batch = unseen.slice(0, maxPagesPerRun);
  if (batch.length === 0) {
    console.log("nothing new to collect, exiting");
    return;
  }
  console.log(`collecting ${batch.length} posting(s) this run`);

  const browser = await chromium.launch();
  const page = await browser.newPage();
  const postings: CollectedPosting[] = [];
  // IDs resolved either way (collected or confirmed-expired) — safe to mark
  // seen, since revisiting either wastes a page load for no gain. IDs that
  // errored transiently (timeout, network blip) are deliberately excluded:
  // they stay unseen so the next run retries them from the same candidate
  // list, rather than being silently dropped forever.
  const resolvedIds = new Set<string>();

  try {
    for (let i = 0; i < batch.length; i++) {
      if (i > 0) await sleep(requestIntervalMs);
      const candidate = batch[i]!;
      try {
        const result = await collectOne(page, candidate);
        resolvedIds.add(candidate.id);
        if (result) postings.push(result);
      } catch (error) {
        console.warn(`WARNING: ${candidate.url} failed: ${String(error)}`);
      }
    }
  } finally {
    await browser.close();
  }

  const expired = resolvedIds.size - postings.length;
  console.log(
    `done: ${postings.length} collected, ${expired} expired/redirected, ` +
      `${batch.length - resolvedIds.size} errored (will retry next run)`,
  );

  for (const id of resolvedIds) seenIds.add(id);
  saveSeenIds(seenIdsPath, seenIds);

  if (postings.length === 0) {
    console.log("nothing to ingest, exiting");
    return;
  }

  const body = {
    source: "catho",
    postings: postings.map((p) => ({
      sourceId: p.sourceId,
      payload: p.payload,
    })),
  };
  const response = await fetch(`${apiUrl}/runs/collect/external`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify(body),
  });
  console.log(`ingest: HTTP ${response.status}`);
  console.log((await response.text()).slice(0, 2000));
  if (!response.ok) process.exit(1);
}

void main();
